content/advanced/fine-grained-reactivity.mdx

```diff
@@ -0,0 +1,222 @@
+
+* Create a global currentSubscriber that can keep track of the function we pass to createEffect
+* Register the function we pass to createEffect as the current subscriber
+* When we access a signal, add the current listener to a list of subscribers
+* When we set the signal to a new value, run all subscribers
+
+```js=
+let currentSubscriber = null;
+
+function createSignal(initialValue) {
+  let value = initialValue;
+  // Maintain a list of a signal's own subscribers
+  const subscribers = new Set();
+    
+  function getter() {
+    // Add to subscriber list
+    if (currentSubscriber) {
+      subscribers.add(currentSubscriber);
+    }
+    return value;
+  }
+    
+  function setter(newValue) {
+    value = newValue;
+    // Notify all subscribers of the value change
+    for (const subscriber of subscribers) {
+      subscriber();
+    }
+  }
+  return [getter, setter];
+}
+
+function createEffect(fn) {
+  // Add function as subscriber in global scope
+  currentSubscriber = fn;
+  // Run function to trigger the accessor of any signals that are called
+  fn();
+  // Remove the function as the current subscriber
+  currentSubscriber = null;
+}
+```
+
+This is all the code needed to create a basic reactive system. We can demonstrate that it works by incrementing the `count` value every second and watching the console.
+
+```js=
+const [count, setCount] = createSignal(0);
+
+createEffect(() => {
+  console.log("The count is " + count());
+});
+
+setInterval(() => {
+  setCount(count() + 1);
+}, 1000);
+```
+
+## Lifecycles in a reactive system
+
+Reactive systems maintain connections between many of its reactive nodes. Meaning, whenever any node is changed parts connected to that node are re-evaluated leading to the update, creation or removal of connections.
+
+Consider when a condition determines the data used to derive a value:
+
+```js
+console.log("1. Create");
+const [firstName, setFirstName] = createSignal("John");
+const [lastName, setLastName] = createSignal("Smith");
+const [showFullName, setShowFullName] = createSignal(true);
+
+const displayName = createMemo(() => {
+  if (!showFullName()) return firstName();
+  return `${firstName()} ${lastName()}`
+});
+
+createEffect(() => console.log("My name is", displayName()));
+
+console.log("2. Set showFullName: false ");
+setShowFullName(false);
+
+console.log("3. Change lastName");
+setLastName("Legend");
+
+console.log("4. Set showFullName: true");
+setShowFullName(true);
+```
+
+In the code above we're making use of a `createMemo` primitive which caches the state of a computation, meaning that the computation doesn't need to be run again if its dependencies do not change. 
+
+In `displayName` notice that there's an early return when the value of `showFullName` is false. When `showFullName` is false only `firstName` is returned, leading to `lastName` being untracked.
```

Comment:
It might be useful to also introduce this _without_ the memo involved, to show that only dependencies read in the last run are tracked.

Reply:
Personally, I would remove the memo from the example, and then maybe show another example with a memo being used in multiple places. Because in this example, the observed behavior (in the logs) will be the same whether there is a memo there or not.

Admittedly, the memo is present in [Ryan's example](https://dev.to/ryansolid/building-a-reactive-library-from-scratch-1i0p), but I honestly don't know why it's there, see for example [this comment thread there](https://dev.to/bluebill1049/comment/1bjgj) for discussion about the use of `createMemo`, though it doesn't explain why Ryan used it in the article.